### ISSUE TITLE
Stripping level set to 'Strip Byte Code causes "AsyncTask.PerformActionInUIThread()" in ARCoreSession.cs to never execute

### ISSUE THREAD

**Issue by kenny-uken** [google-ar/arcore-unity-sdk#73]
**edit: skip here to find the real issue:**  https://github.com/google-ar/arcore-unity-sdk/issues/73#issuecomment-358441294

I have a Samsung S8 with arcore_preview2.apk installed on it. I am able to build and run the HelloAR demo project successfully.

I have a game which worked with Preview 1 but now refuses to connect the AR session, throwing no errors of any kind either in the Unity logs or in adb logcat. I copied over the "ARCore Device" prefab from HelloAR just to make sure everything was set up correctly and double-checked my Player Settings to make sure they matched, but nothing seems to work.

The SessionConnectionState is always "Uninitialized". When debugging the game it seems like the anonymous function in ARCoreSession.cs line 150 ("returnTask.ThenAction()") which is supposed to set the connection status after the connection attempt finishes is never called.

Here's output from logcat:
```
01-09 15:13:14.566  7054  7183 I Unity   : ArCoreUnity_createSession
01-09 15:13:14.566  7054  7183 I native  : session_create.cc:62 Entering ArSession_create.
01-09 15:13:14.586  7054  7183 I native  : session_create.cc:28 DYNAMITE_LOAD_RESULT_OK
01-09 15:13:14.595  7054  7183 I native  : session_create_implementation.cc:329 Entering ArSession_createImplementation. ARCore SDK version: [0.91.0].
01-09 15:13:14.769  7054  7183 I native  : session_create_implementation.cc:366 ArSession_createImplementation returning OK.
01-09 15:13:14.802  7054  7054 E Unity   :   at GoogleARCoreInternal.NativeApi+ExternApi.ArCoreUnity_resumeSession (GoogleARCoreInternal.EarlyUpdateCallback earlyUpdate) [0x00000] in <filename unknown>:0
01-09 15:13:14.802  7054  7054 E Unity   :   at GoogleARCore.ARCoreSession+<_ResumeSession>c__AnonStorey1.<>m__0 () [0x00000] in <filename unknown>:0
01-09 15:13:14.913  7054  7183 E native  : session_c_api.cc:587 ArCamera_getTrackingState: camera was passed NULL.
01-09 15:13:14.917  7054  7183 E native  : session_c_api.cc:674 ArFrame_getLightEstimate: frame was passed NULL.
01-09 15:13:14.939  7054  7183 E native  : session_c_api.cc:587 ArCamera_getTrackingState: camera was passed NULL.
01-09 15:13:14.939  7054  7183 E native  : session_c_api.cc:587 ArCamera_getTrackingState: camera was passed NULL.
01-09 15:13:14.947  7054  7183 E native  : session_c_api.cc:587 ArCamera_getTrackingState: camera was passed NULL.
01-09 15:13:14.947  7054  7183 E native  : session_c_api.cc:674 ArFrame_getLightEstimate: frame was passed NULL.
```
The last two messages are repeated every frame.

**Comment by kenny-uken** [comment]
actually I unintentionally filtered that logcat output too much and found that the game can't find "unitygar", so I'm guessing it's somehow not including the contents of GoogleARCore/SDK/Plugins/

**Comment by chaosemer** [comment]
Could you please send the full results of `adb bugreport`? Filtering down the logcat makes it harder to diagnose. Thanks.

**Comment by chucknology** [comment]
See if "ARCore supported" is checked in the Unity XR section of your Player Settings.

**Comment by kenny-uken** [comment]
It /definitely/ is, I just checked again now just to be certain. I have even reproduced the gradle build settings in HelloAR that I have set up in my own project and made sure those plus the player settings are identical.

**Comment by chucknology** [comment]
OK. Is "Connect on awake" checked on the object "ARCore Device", component "ARCoreSession"?

**Comment by kenny-uken** [comment]
Yes. I'm using the "ARCore Device" prefab provided in the SDK with a few gameobjects attached which don't touch any ARCore components or code.

**Comment by chucknology** [comment]
Hmm. We had this issue with the update to ARCore Preview 2, but I can't think how we resolved it, if it wasn't any of those suggestions.
I'll post again if it comes to me.
The unitygar exception is still thrown even when it's working, btw.
In the meantime, if you can provide a log as Jared has asked for above, maybe he can help.

**Comment by kenny-uken** [comment]
For the sake of everyone reading this, I sent the bugreport privately to him already :)

**Comment by kenny-uken** [comment]
I just checked HelloAR again, and it indeed does throw the `Unable to find unitygar` error, so I guess that's meaningless.

**Comment by kenny-uken** [comment]
One of my colleagues found the core issue and we're up and running again!

https://github.com/google-ar/arcore-unity-sdk/blob/master/Assets/GoogleARCore/SDK/Scripts/ARCoreSession.cs#L203

It seems that this AsyncTask was never being run for our project. We removed the "AsyncTask.PerformActionInUIThread()" wrapper and just ran the contained code as-is. This resolved our issue AND does not seem to break the pause / resume functionality which the comment above that line suggests might be broken if it were not an AsyncTask.

**Comment by fredsa** [comment]
In that ARCoreSession.cs you can see that AsyncTask relies on `Update()` being called. Please check that the script component is not disabled, and that the game object it belongs to is also not disabled. That includes any game object ancestors.

**Comment by kenny-uken** [comment]
It seems that "OnUIUpdate();" is never getting called by s_CallOnUIThread in AsyncTask's constructor. The code that the SDK wants to run in that AsyncTask just sits around in s_UiThreadActionQueue forever. I believe it's because we're using Fabric for crash reporting, which wraps the Unity player, so " AndroidJavaClass unityPlayerClass = new AndroidJavaClass("com.unity3d.player.UnityPlayer"); " might not be finding the right process. I'm not having much luck changing that class name to "io.fabric.unity.android.FabricApplication" though. This is a bit out of my expertise so maybe I'm looking in the wrong place.

**Comment by chaosemer** [comment]
Unfortunately, I'm not familiar with Fabric. The key part is that the activity that gets "runOnUiThread" called needs to be the active activity.

**Comment by chucknology** [comment]
Historically, Google Unity SDKs have used a "GoogleUnityActivity" that extends the UnityPlayerActivity for hooking the Android lifecycle independently of Unity.
If the Fabric activity is extending the base UnityPlayer instead of the GoogleUnityActivity, then the Activity in use will not have the Google lifecycle functions.
This design is changing with ARCore and the associated versions of Unity.
Integrating Fabric by subclassing the Application as opposed to modifying the Activity, is a more decoupled approach that is better suited for Unity development.

**Comment by chaosemer** [comment]
We're not using GoogleUnityActivity any more; one of the benefits of native integration is we can now get the events we wanted from Unity w/o writing our own activity.

**Comment by kenny-uken** [comment]
A colleague of mine figured out a very simple solution that will allow us to undo our workaround. Here's a description of the fix in his own words:

There was an exception `no such proxy method: UnityEngine.AndroidJavaRunnableProxy`

In our case this method was stripped, we had stripping level set to 'Strip Byte Code'. So we can change this OR add one line to link.xml to never strip this specific method.

To add it, we need to add a line to file Assets/link.xml:
`<type fullname="UnityEngine.AndroidJavaRunnableProxy" preserve="all"/>`

If this file doesn't exist, then create a new one with contents:
```
<linker>
         <assembly fullname="UnityEngine">
                 <type fullname="UnityEngine.AndroidJavaRunnableProxy" preserve="all"/>
         </assembly>
</linker>
```

**Comment by chaosemer** [comment]
This has been fixed in ARCore 1.0.
